lidar_pipeline:
  ros__parameters:
    hesai:
      parameters:
        x_min: -20.0       # Dufour: front, Castor: left
        x_max: -0.1         # Dufour: rear,  Castor: right
        y_min: -10.0       # Dufour: right, Castor: front
        y_max: 10.0        # Dufour: left,  Castor: rear
        z_min: -2.0        # Dufour: down,  Castor: down
        z_max: 0.5         # Dufour: up,    Castor: up
        x_min_car: -3.0    # Dufour: front, Castor: left
        x_max_car: 0.0     # Dufour: rear,  Castor: right
        y_min_car: -1.0    # Dufour: right, Castor: front
        y_max_car: 1.0     # Dufour: left,  Castor: rear

        # x_min: -8.0       # Dufour: front, Castor: left
        # x_max: 8.0         # Dufour: rear,  Castor: right
        # y_min: -25.0       # Dufour: right, Castor: front
        # y_max: 0.0        # Dufour: left,  Castor: rear
        # z_min: -2.0        # Dufour: down,  Castor: down
        # z_max: 1.5         # Dufour: up,    Castor: up
        # x_min_car: -3.0    # Dufour: front, Castor: left
        # x_max_car: 0.0     # Dufour: rear,  Castor: right
        # y_min_car: -1.0    # Dufour: right, Castor: front
        # y_max_car: 1.0     # Dufour: left,  Castor: rear

        grid_num_x: 4
        grid_num_y: 4
        distance_threshold_ransac_near: 0.12
        distance_threshold_ransac_far: 0.12
        distance_threshold_far_near: 12.0


        # in skidpad, if missing a lot of cones, try to decrease this, or just set to 0.0
        # in autox / trackdrive: 0.89
        min_fraction_inlier: 0.7
        min_point_in_segment: 150

        rain: false


        grid_filter_enabled: true                      # lidar_only: false, sensor_fusion: true
        rough_clustering_max_points_per_grid: 120
        rough_clustering_grid_size_m: 1.5

        cone_detection:
          enabled: true
          max_points_dbscan: 300
          max_cluster_z: 0.5
          min_cluster_std: 0.04
          min_z_size: 0.08
          zones:
            x: [9.0, 10.0, 50.0]
            y: [4.0, 5.0, 50.0]

            # eps_dbscan_m:
            # skidpad: [1.5, 1.5, 1.5]
            # autox and trackdrive:
            #   lidar_only: [0.1, 1.0, 1.8]
            #   sensor_fusion: [1.8, 1.8, 1.8]
            eps_dbscan_m: [1.5, 1.5, 1.5]
            min_points_dbscan: [4, 4, 4]
            std_threshold_orange: [0.10, 0.50, 0.50]
            max_cluster_std_dev: [0.15, 0.12, 0.12]

        pre_cluster:
          enabled: false
          max_points_dbscan: 300
          max_cluster_z: -0.5
          min_cluster_std: 0.05
          min_z_size: 0.1
          zones:
            x: [10.0, 50.0]
            y: [4.0, 50.0]
            eps_dbscan_m: [0.1, 1.5]
            min_points_dbscan: [10, 4]
            std_threshold_orange: [0.10, 0.20]
            max_cluster_std_dev: [0.13, 0.135]

      profiling:
        enabled: true

      logging:
        enabled: true

      visualization:
        enabled: true

      filter:
        enabled: true

      motion_compenation:
        enabled: true

      transform_cone:
        enabled: true

      debug:
        enabled: false

      topics:
        subscriber:
          topic_lidar_point: "/lidar_points"
        publisher:
          topic_cone_array: "/perception/lidar/cone_array"
          topic_compensated_pc: "/perception/lidar_motion_compensator/compensated_pc"
          topic_debug_filtered_pc: "/filtered_point_cloud"
          topic_debug_centroid_dbscan: "/cluster_centroids_markers"

      frame_id:
        fixed: "odom"
        cone_array: "base_link"

      ground_removal:
        max_iteration: 50

      downsampling:
        ratio: 1.0
        enabled: false
